Framework/DataHandling/src/SaveNXSPE.cpp

```diff
@@ -115,15 +116,55 @@ void SaveNXSPE::exec() {
 
   // Get the value out of the property first
   double efixed = getProperty("Efixed");
-  if (isEmpty(efixed))
+  bool efixed_provided(true);
+  if (isEmpty(efixed)) {
     efixed = MASK_FLAG;
-  // Now lets check to see if the workspace knows better.
-  // TODO: Check that this is the way round we want to do it.
-  const API::Run &run = inputWS->run();
-  if (run.hasProperty("Ei")) {
-    efixed = run.getPropertyValueAsType<double>("Ei");
+    efixed_provided = false;
   }
-  nxFile.writeData("fixed_energy", efixed);
+  auto eMode = inputWS->getEMode();
+  bool write_single_energy(true);
+  if (!efixed_provided) {
+    // efixed identified differently for different types of inelastic instruments
+    // Now lets check to see if we can retrieve energy from workspace.
+    switch (eMode) {
+    case (Kernel::DeltaEMode::Indirect): {
+      // here workspace detectors should always have the energy
+      auto allEi = getIndirectEfixed(inputWS);
+      if (allEi.size() == 1) {
+        efixed = allEi[0];
+      } else if (allEi.size() > 1) {
+        write_single_energy = false; // do not write single energy, write array of energies here
+        nxFile.writeData("fixed_energy", allEi);
+      }
+      if (!write_single_energy || (allEi[0] > std::numeric_limits<float>::epsilon()))
+        // this is generally incorrect, but following historical practice
+        break; // assume that indirect instrument without energy attached to detectors
+               // may have energy specified in Ei log. Some user scripts may depend on it.
+               // so here we break only if Ei is defined on detectors and look for
+               // Ei log otherwise
+    }
+    case (Kernel::DeltaEMode::Elastic):   // no efixed for elastic,
+                                          // whatever retrieved from the property should remain unchanged.
+                                          // Despite it is incorrect, previous tests were often using Instrument in
+                                          // Elastic mode as the source of data for Direct inelastic. Despite correct
+                                          // action would be no efixed for elastic, to keep tests happy here we try to
+                                          // derive efixed from Ei log similarly to Direct inelastic case if no external
+                                          // efixed provided to the algorithm.
+    case (Kernel::DeltaEMode::Undefined): // This should not happen
+                                          // but to keep cpp-check happy, assume Direct
```

Comment:
Would disagree with this change. 

this file format does not make sence for real elastic instruments. We do not fail it for Elastic only because were lazy to modify test instruments into appropriate mode.  
Number of tests have instrument set to Elastic but add Ei log and test nxspe as for direct (or indirect). In this case, data will be written as Direct (now we write Direct and Indirect instruments differently), so we assume Direct.

Reply:
OK I understand, because both `Elastic` and `Undefined` are assumed to be `Direct` instrument with Ei in the log. From the comment it looked like you would expect the `Undefined` case to run the same code as the `Direct` - whereas in fact it has the same result as `Elastic`. If it's clear to you then that's fine with me!

Reply:
for `Elastic`, we lazily run code for `Direct`. I will probably add the change of mode too to write `NXfermi` below